spec-tests/src/test/java/org/candlepin/spec/OwnerProductResourceSpecTest.java

```diff
@@ -0,0 +1,799 @@
+        String originalProductUuid = product.getUuid();
+        product = ownerProductApi.updateProductByOwner(ownerKey, product.getId(), product);
+
+        assertEquals(1, product.getProvidedProducts().size());
+        ProductDTO providedProduct = Iterables.getOnlyElement(product.getProvidedProducts());
+        assertEquals(newProvidedProduct, providedProduct);
+
+        // A new product version should have been created during an update of provided product
+        assertNotEquals(originalProductUuid, product.getUuid());
+    }
+
+    @Test
+    @DisplayName("should allow deleting a product associated with an activation key")
+    public void shouldAllowDeletingAProductAssociatedWithAnActivationKey() throws Exception {
+        OwnerDTO owner = ownerApi.createOwner(Owners.random());
+        String ownerKey = owner.getKey();
+        ProductDTO product = ownerProductApi.createProductByOwner(ownerKey, Products.random());
+        ActivationKeyDTO activationKey = ownerApi.createActivationKey(ownerKey, ActivationKeys.random(owner));
+        activationKeyApi.addProductIdToKey(activationKey.getId(), product.getId());
+
+        // The activation key should have the product associated with it
+        ActivationKeyDTO actual = activationKeyApi.getActivationKey(activationKey.getId());
+        assertNotNull(actual);
+        List<ActivationKeyProductDTO> akProducts = actual.getProducts().stream()
+            .filter(p -> p.getProductId().equals(product.getId()))
+            .collect(Collectors.toList());
+        assertEquals(1, akProducts.size());
+
+        // Deleting the product should remove its reference from the activation key
+        ownerProductApi.deleteProductByOwner(ownerKey, product.getId());
+
+        // The activation key should still exist, but should no longer reference the product
+        actual = activationKeyApi.getActivationKey(activationKey.getId());
+        assertNotNull(actual);
+        assertEquals(0, actual.getProducts().size());
+    }
+
+    @Test
+    @DisplayName("should return bad request when inserting product with null name")
+    public void shouldReturnBadRequestWhenInsertingProductWithNullName() throws Exception {
+        OwnerDTO owner = ownerApi.createOwner(Owners.random());
+        ProductDTO product = Products.random()
+            .name(null);
+
+        Request request = Request.from(client)
+            .setPath(OWNER_PRODUCTS_PATH)
+            .setMethod("POST")
+            .setPathParam("owner_key", owner.getKey())
+            .setBody(product);
+
+        assertEquals(400, request.execute().getCode());
+    }
+
+    @Test
+    @DisplayName("should return bad request when inserting product with null id")
+    public void shouldReturnBadRequestWhenInsertingProductWithNullId() throws Exception {
+        OwnerDTO owner = ownerApi.createOwner(Owners.random());
+        ProductDTO product = Products.random()
+            .id(null);
+
+        Request request = Request.from(client)
```

Comment:
What is the `createProductByOwner` endpoint lacking that these two tests need?

Reply:
Created the JsonManipulationTests nested test class to have tests that serialize nulls, and then a set of tests that do no serialize the null values.